// roulette-spinner/dist/builders/board.js
import { toHTMLBoard } from '../utils/utilities';
export class BoardBuilder {
    element;
    custom_image;
    doughnut;
    shift;
    border;
    radius;
    padding;
    constructor(board) {
        this.shift = board.shift;
        this.border = board.border;
        this.radius = board.radius;
        this.padding = board.padding;
        this.doughnut = board.doughnut;
        if (board.element) {
            this.element = toHTMLBoard(board.element);
            this.element.style.width = `${this.radius * 2 - this.padding * 2}px`;
            this.element.style.height = `${this.radius * 2 - this.padding * 2}px`;
            this.element.style.padding = `${this.padding}px`;
        }
        else {
            this.element = this.createSVGBoard();
        }
        this.custom_image = !!board.element;
    }
    rotateImage(degrees) {
        this.element.style.transform = `rotate(${degrees}deg)`;
    }
    createSVGBoard() {
        const svg = document.createElementNS('http://www.w3.org/2000/svg', 'svg');
        svg.setAttribute('width', `${this.radius * 2}`);
        svg.setAttribute('height', `${this.radius * 2}`);
        svg.setAttribute('viewBox', `0 0 ${this.radius * 2} ${this.radius * 2}`);
        svg.setAttribute('id', 'roulette-circle');
        svg.style.transform = `rotate(${this.shift * -1}deg)`;
        return svg;
    }
}
