// @theblindhawk/roulette/dist/builders/arrow.js
import { getShiftValue, toHTMLArrow } from '../utils/utilities';
export class ArrowBuilder {
    element;
    width;
    height;
    padding;
    color;
    shift;
    constructor(arrow) {
        this.element = toHTMLArrow(arrow.element);
        this.width = arrow.width;
        this.height = arrow.height;
        this.padding = arrow.padding;
        this.color = arrow.color;
        this.shift = getShiftValue(arrow.shift);
        this.drawArrow();
    }
    drawArrow() {
        this.element.style.width = `${this.width}px`;
        this.element.style.height = `${this.height}px`;
        this.element.style.position = 'absolute';
        this.element.style.margin = '0 auto auto auto';
        this.element.style.top = `0px`;
        this.element.style.left = `0px`;
        this.element.style.right = `0px`;
        this.element.style.bottom = `0px`;
        this.element.setAttribute('fill', this.color);
        this.element.style.transform = `rotate(${this.shift}deg)`;
        this.element.style.zIndex = '1';
    }
}
